zookeeper/kv_zookeeper.go

```diff
@@ -0,0 +1,629 @@
+		time.Sleep(time.Second)
+		kvPair, err = z.Create(key, lockTag, 1)
+		if count > 0 && count%2 == 0 && err != nil {
+			currLockerTag := LockerIDInfo{}
+			if _, errGet := z.GetVal(key, &currLockerTag); errGet == nil {
+				logrus.Warnf("Lock %v locked for %v seconds, tag: %v, err: %v",
+					key, count, currLockerTag, err)
+			}
+		}
+		if err != nil && time.Since(startTime) > lockTryDuration {
+			return nil, err
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	kvPair.Lock = &zookeeperLock{Done: make(chan struct{})}
+	go z.waitForUnlock(kvPair, lockerID, lockHoldDuration)
+	return kvPair, err
+}
+
+func (z *zookeeperKV) waitForUnlock(
+	kvp *kvdb.KVPair,
+	lockerID string,
+	lockHoldDuration time.Duration,
+) {
+	l := kvp.Lock.(*zookeeperLock)
+	timer := time.NewTimer(lockHoldDuration)
+	defer timer.Stop()
+
+	lockMsgString := kvp.Key + ",tag=" + lockerID
+
+	for {
+		select {
+		case <-timer.C:
+			if lockHoldDuration > 0 {
+				l.Lock()
+				defer l.Unlock()
+				if !l.Unlocked {
+					if _, err := z.Delete(kvp.Key); err != nil {
+						logrus.Errorf("Error deleting lock %s after timeout. %v",
+							lockMsgString, err)
+					}
+					l.Unlocked = true
+				}
+				z.FatalCb("Lock %s hold timeout triggered", lockMsgString)
+				return
+			}
+		case <-l.Done:
+			return
+		}
+	}
+}
+
+func (z *zookeeperKV) Unlock(kvp *kvdb.KVPair) error {
+	l, ok := kvp.Lock.(*zookeeperLock)
+	if !ok {
+		return fmt.Errorf("Invalid lock structure for key: %s", kvp.Key)
+	}
+	l.Lock()
+	if _, err := z.Delete(kvp.Key); err != nil {
+		l.Unlock()
+		return err
+	}
+	l.Unlocked = true
+	l.Unlock()
+	l.Done <- struct{}{}
+	return nil
+}
+
+func (z *zookeeperKV) TxNew() (kvdb.Tx, error) {
+	return nil, kvdb.ErrNotSupported
+}
+
+func (z *zookeeperKV) AddUser(username string, password string) error {
+	return kvdb.ErrNotSupported
+}
+
+func (z *zookeeperKV) RemoveUser(username string) error {
+	return kvdb.ErrNotSupported
+}
+
+func (z *zookeeperKV) GrantUserAccess(
+	username string,
+	permType kvdb.PermissionType,
+	subtree string,
+) error {
+	return kvdb.ErrNotSupported
+}
+
+func (z *zookeeperKV) RevokeUsersAccess(
+	username string,
+	permType kvdb.PermissionType,
+	subtree string,
+) error {
+	return kvdb.ErrNotSupported
+}
+
+func (z *zookeeperKV) Serialize() ([]byte, error) {
+	return nil, kvdb.ErrNotSupported
+}
+
+func (z *zookeeperKV) Deserialize(b []byte) (kvdb.KVPairs, error) {
+	return nil, kvdb.ErrNotSupported
+}
+
+// normalize converts a given path to the form /a/b/c
+func normalize(key string) string {
+	if key == "" {
+		return ""
+	}
+	path := strings.Split(strings.Trim(key, "/"), "/")
+	return "/" + strings.Join(path, "/")
+}
+
+// createFullPath creates the entire path for a directory
+func (z *zookeeperKV) createFullPath(key string, ephemeral bool) error {
+	key = z.domain + normalize(key)
+	path := strings.Split(strings.TrimPrefix(key, "/"), "/")
+
+	for i := 1; i <= len(path); i++ {
+		newPath := "/" + strings.Join(path[:i], "/")
```

Comment:
There is no prefix `/` because we remove it when we split the path in array.
```
path = [pwx,cluster,database]
iteration 1 = / + pwx
iteration 2 = / + join([pwx,cluster], /) = / + pwx/cluster
```